### ISSUE TITLE
Installed PyEnchant but Manuskript still asks me to "Install PyEnchant to use Spellcheck"

### ISSUE THREAD

**Issue by Weej1** [olivierkes/manuskript#122]
Hi! Been keeping an eye on this since the first version 
I've finally started using it for a technical document and I tried to get PyEnchant to work but even though I installed it (pip install pyenchant) I still get the message to install PyEnchant in the menu. 

Also is there a way to insert images? if yes how?

**Comment by gedakc** [comment]
What Operating System are you using?

There are installation steps for GNU/Linux, Windows, and OS X on the [wiki](https://github.com/olivierkes/manuskript/wiki).

To insert images, upload the image to a file sharing site, and then in the post add a link to the image.

**Comment by Weej1** [comment]
HI gedakc - Thanks for the quick reply! 
I'm using Lubuntu 17.04 and I did install both the python3 and the standard pip variation of pyenchant. But I still the message in tools. 

Also for the image - will it automatically download from the site and insert into the document when I compile ?

**Comment by gedakc** [comment]
I recall reading that using both debian packages and pip might confuse a python installation.

I assume that _python3_ was installed as a debian package.  As such I suggest installing _pyenchant_ with the following command:

`sudo apt install python3-enchant`

See also [Run Manuskript from Source Code on Linux](https://github.com/olivierkes/manuskript/wiki/Run-Manuskript-from-Source-Code-on-Linux).

> Also for the image - will it automatically download from the site and insert into the document when I compile ?

Yes for images you wish to appear in this github issue tracker.

For images within a manuskript document see issue #22.

**Comment by ScullyBlue** [comment]
I have the same problem.

Output, when I start from terminal:
Running manuskript version 0.4.0.
Warning: failed to load translator for locale de_DE...
Unable to update the static FcBlanks: 0x0600
Unable to update the static FcBlanks: 0x0601
Unable to update the static FcBlanks: 0x0602
Unable to update the static FcBlanks: 0x0603
Unable to update the static FcBlanks: 0x06dd
Unable to update the static FcBlanks: 0x070f
Unable to update the static FcBlanks: 0x2028
Unable to update the static FcBlanks: 0x2029
Unable to update the static FcBlanks: 0xfff9
Unable to update the static FcBlanks: 0xfffa
Unable to update the static FcBlanks: 0xfffb

And the GUI says, that I need to install pyenchant, to use spellchecker. Version 3.0.0 works instead of this fine.
I uninstall the enchant-module and install it via pip. It also not working. Then I saw, that python 3 is requiered, so I install the python3-enchant package (via APT and after that via pip3) and it also doesnt work. But Debian link the command "python" default to python2.7, not to python3.4. So I change the shebang of the manuscript-python-scripts from python to python3, that links to the newest python3 thats installed on computer. It works also not. 

In short:
By default setup (Debian 8.8, python2.7 + pyenchant 1.6.8 / python3 + pyenchant 1.6.8 from standard)  - dont work for me
Setup via pip/pip3 pyenchant - dont work for me
Reinstall via APT python-echant/python3-enchant - dont work for me
Change symlink for python from python2.7 to python3.4 - dont work for me
Change python-shebang in .py-Scripts - dont work for me

I dont have any ideas ... Oh, and I saw, that manuscript brings libs for python3.5 ... is that python-version requiered?

**Comment by gedakc** [comment]
@ScullyBlue, from your description I assume you are configuring manuskript to run from source code.

Have you tried the steps listed in [Run Manuskript from Source Code on Linux](https://github.com/olivierkes/manuskript/wiki/Run-Manuskript-from-Source-Code-on-Linux)?

**Comment by ScullyBlue** [comment]
Yes, I use the ZIP from homepage, unzip it in my home-directory and prove/changed owner/group
I installed also all dependend packages. As I said, under Debian 8.8 all python3 packages are in the python3.4-Version, unter Debian 9 (updated my system) its Python3.5 (doesnt work too ...) - beside python-packages, which are in Version 2.7 and called via command "python" instead of "python3".

# python -V
Python 2.7.13

# python3 -V
Python 3.5.3

# pip3 list --format=legacy
[...]
lxml (3.7.1)
Markdown (2.6.8)
[...]
pyenchant (1.6.7)
[...]

# dpkg -l | grep <packages>
[...]
ii/**libqt5svg5:amd64**/5.7.1~20161021-2+b2/amd64/Qt/5/SVG/module
[...]
ii/**python3**/3.5.3-1/amd64/interactive/high-level/object-oriented/language/(default/python3/version)
ii/**python3-enchant**/1.6.7-1/all/spellchecking/library/for/Python/3
ii/**python3-lxml**/3.7.1-1/amd64/pythonic/binding/for/the/libxml2/and/libxslt/libraries
ii/**python3-markdown**/2.6.8-1/all/text-to-HTML/conversion/library/tool/(Python/3/version)
ii/**python3-pyqt5**/5.7+dfsg-5/amd64/Python/3/bindings/for/Qt5
ii/**python3-pyqt5.qtwebkit**/5.7+dfsg-5/amd64/Python/3/bindings/for/Qt5's/WebKit/module
[...]
ii/**zlib1g:amd64**/1:1.2.8.dfsg-5/amd64/compression/library/-/runtime

**Comment by gedakc** [comment]
> Yes, I use the ZIP from homepage...

Which zip file did you use?
- **manuskript-0.4.0-linux.zip** file under _Current Version (0.4.0)_ then _Linux 64 bits_.
   if this file is used then it already contains Python3, PyQt5 and many other dependencies.  No additional packages are required.  If PyEnchant is missing from this install then it is possible that the zip file was created without PyEnchant.  Installing PyEnchant from pip or a package repository will not affect this self-contained package.
-  **manuskript-0.4.0.zip** (listed as **zip**) file under _Sources_ then _Version 0.4.0_.
   If this file is used then all the dependencies must be installed as per the steps described in [Run Manuskript from Source Code on Linux](https://github.com/olivierkes/manuskript/wiki/Run-Manuskript-from-Source-Code-on-Linux).

**Comment by ScullyBlue** [comment]
Oh, sorry, my fault: I used **manuskript-0.4.0-linux.zip**.
Give me a moment and I test it with the source-ZIP.

**Comment by gedakc** [comment]
@ScullyBlue I just installed `manuskript-0.4.0-linux.zip` (self contained package) on 64 bit Kubuntu 16.04.  I can confirm that it runs but the **Tools** menu indicates _Install PyEnchant to use spellcheck_.

@olivierkes perhaps this package was built without PyEnchant?
If so we should rebuild and republish the self contained `manuskript-0.4.0-linux.zip` package.

**Comment by ScullyBlue** [comment]
@gedakc Yes, the Software runs (and other aspects work very nicely! (: ), but the Spellcheck dont work and give the same message. "Install PyEnchant to use spellcheck."

The Source-Installation crashes by opening a projekt.

Traceback (most recent call last):
  File "/opt/manuskript-0.4.0/bin/../manuskript/ui/welcome.py", line 119, in loadRecentFile
    self.mw.loadProject(act.data())
  File "/opt/manuskript-0.4.0/bin/../manuskript/mainWindow.py", line 322, in loadProject
    self.toggleSpellcheck(settings.spellcheck)
  File "/opt/manuskript-0.4.0/bin/../manuskript/mainWindow.py", line 1003, in toggleSpellcheck
    w.toggleSpellcheck(val)
  File "/opt/manuskript-0.4.0/bin/../manuskript/ui/views/textEditView.py", line 360, in toggleSpellcheck
    self._dict = enchant.Dict(self.currentDict)
  File "/usr/lib/python3/dist-packages/enchant/__init__.py", line 558, in __init__
    _EnchantObject.__init__(self)
  File "/usr/lib/python3/dist-packages/enchant/__init__.py", line 168, in __init__
    self._init_this()
  File "/usr/lib/python3/dist-packages/enchant/__init__.py", line 565, in _init_this
    this = self._broker._request_dict_data(self.tag)
  File "/usr/lib/python3/dist-packages/enchant/__init__.py", line 310, in _request_dict_data
    self._raise_error(eStr % (tag,),DictNotFoundError)
  File "/usr/lib/python3/dist-packages/enchant/__init__.py", line 258, in _raise_error
    raise eclass(default)
enchant.errors.DictNotFoundError: Dictionary for language 'de_DE' could not be found
Fatal Python error: Aborted

Current thread 0x00007fb654622700 (most recent call first):
  File "/opt/manuskript-0.4.0/bin/../manuskript/main.py", line 73 in launch
  File "/opt/manuskript-0.4.0/bin/../manuskript/main.py", line 64 in run
  File "./bin/manuskript", line 13 in <module>

**Comment by gedakc** [comment]
> enchant.errors.DictNotFoundError: Dictionary for language 'de_DE' could not be found
> Fatal Python error: Aborted

It seems that Manuskript cannot find the de_DE dictionary.

On my Kubuntu 16.04 I noticed that it also has some other packages installed that contain the word `enchant`.  Can you check if your installation has these packages?

```
$ dpkg -l | grep enchant
ii  enchant                                         1.6.0-10.1build2                           amd64        Wrapper for various spell checker engines (binary programs)
ii  libenchant1c2a:amd64                            1.6.0-10.1build2                           amd64        Wrapper library for various spell checker engines (runtime libs)
ii  python3-enchant                                 1.6.6-2                                    all          spellchecking library for Python 3
```

Also it appears that the `libenchant1c2a` package requires `aspell` packages.  Can you check for the existence of these packages?

```
$ dpkg -l | grep aspell
ii  aspell                                          0.60.7~20110707-3build1                    amd64        GNU Aspell spell-checker
ii  aspell-en                                       7.1-0-1.1                                  all          English dictionary for GNU Aspell
ii  libaspell15:amd64                               0.60.7~20110707-3build1                    amd64        GNU Aspell spell-checker runtime library
```

In your situation I expect that `aspell-en` would be `aspell-de`.

**Comment by ScullyBlue** [comment]
I have the same echant-Packages on my debian-system
`$ dpkg -l | grep enchant
ii  enchant                                       1.6.0-11+b1                            amd64        Wrapper for various spell checker engines (binary programs)
ii  libenchant1c2a:amd64                          1.6.0-11+b1                            amd64        Wrapper library for various spell checker engines (runtime libs)
ii  python3-enchant                               1.6.7-1                                all          spellchecking library for Python 3`

But also the same Aspell-Packages - only en, not de.
`$ dpkg -l | grep aspell
ii  aspell                                        0.60.7~20110707-3+b2                   amd64        GNU Aspell spell-checker
ii  aspell-en                                     2016.11.20-0-0.1                       all          English dictionary for GNU Aspell
ii  libaspell15:amd64                             0.60.7~20110707-3+b2                   amd64        GNU Aspell spell-checker runtime library`

After install the german aspell-package, it works fine with the source-installation! Thanks for help! Maybe for the future it would be nice, if the aspell-packages mentioned by dependencies?

**Comment by gedakc** [comment]
Thank you @ScullyBlue for testing and confirming that the `aspell-de` German language dictionary was required.

As per your suggestion I have updated the PyEnchant section of the [Run Manuskript from Source Code on Linux](https://github.com/olivierkes/manuskript/wiki/Run-Manuskript-from-Source-Code-on-Linux) instructions to include the desired language dictionary.

**Comment by gedakc** [comment]
@olivierkes perhaps the `manuskript-0.4.0-linux.zip` package was built without _PyEnchant_ or the _aspell-xx_ language dictionaries?

If so we should rebuild and republish the self contained `manuskript-0.4.0-linux.zip` package.

I confirmed the issue outlined in this report.  I installed `manuskript-0.4.0-linux.zip` (self contained package) on 64 bit Kubuntu 16.04.  I can run manuskipt from the extracted zip file; however the **Tools** menu indicates to _Install PyEnchant to use spellcheck_.

I have also tested the `manuskript-0.4.0-win32.zip` (self contained package) for Windows.  This windows zip file does contain the following spellcheck dictionaries:  `de_DE, en_AU, en_GB, en_US, and fr_FR`.

**Comment by gedakc** [comment]
It appears that _PyInstaller_, which we currently use to create our Linux and Windows packages, supports _PyEnchant_ spell checking on Windows only.  See [PyInstaller - Supported Packages](https://github.com/pyinstaller/pyinstaller/wiki/Supported-Packages) and search for `pyenchant`.

I can  confirm that PyEnchant is not working on Linux with the `manuskript-0.4.0-linux.zip` package.

I will continue to investigate packaging options for manuskript on Linux and Windows.
I don't have a Mac OS X computer.  Any help with Mac packaging is appreciated.

**Comment by gedakc** [comment]
@Weej1 and @ScullyBlue, would you be able to try a new manuskript package I created?

I created a mirror site on SourceForge and uploaded the package there.

Download [manuskript-0.4.0-1-linux64.zip](https://sourceforge.net/projects/manuskript/files/manuskript-0.4.0/manuskript-0.4.0-1-linux64.zip/download).

This version should enable spell check to work on several GNU/Linux distributions, with the exception of Fedora and Debian 9.

Be sure to install a dictionary for your language.  For example for French use:
```
sudo apt install aspell-fr
```

Please let me know if this new package enables spell check for you.

Thanks in advance.

**Comment by Weej1** [comment]
Hi @gedakc 
Can confirm `Segmentation fault (core dumped)` on my end. 
I'm running elementaryOS loki (0.4.1) and using aspell-en (was pre-installed)
Tried moving the `libstdc++.so.6` file too on a whim, it didn't fix the crash though. 
The standard manuskript 0.4.0 linux version works fine though. 

Anything else I can try to fix this? Will be happy to help.

**Comment by gedakc** [comment]
Thanks @Weej1 for trying the download and reporting back that it caused a segmentation fault on your elementaryOS system.

Following is a quick analysis of the core dump I had on Debian 9:

```
user@debian9:~$ cd manuskript/
user@debian9:~/manuskript$ ulimit -c unlimited
user@debian9:~/manuskript$ ./manuskript 
Segmentation fault (core dumped)
user@debian9:~/manuskript$ gdb ./manuskript core 
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./manuskript...(no debugging symbols found)...done.
[New LWP 1962]
Core was generated by `./manuskript'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __GI_raise (sig=11) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=11) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x0000000000404c5c in ?? ()
#2  0x0000000000402d6d in ?? ()
#3  0x00007f799578d2b1 in __libc_start_main (main=0x4018f0, argc=1, 
    argv=0x7fff18823848, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fff18823838)
    at ../csu/libc-start.c:291
#4  0x000000000040191e in ?? ()
(gdb) quit
user@debian9:~/manuskript$ 
```

I will continue to investigate.  I am open to suggestions on how to resolve the issue.

**Comment by Spirou4D** [comment]
Hi 

> I have also tested the manuskript-0.4.0-win32.zip (self contained package) for Windows. This windows zip file does contain the following spellcheck dictionaries: de_DE, en_AU, en_GB, en_US, and fr_FR.

I installed this on my Windows 7 64b and installed PyEnchant for my python 3,5 for 64b but no yet french translation!
Always this traceback on Teminal console:
Failed to load translator for fr_FR...
and
"Install PyEnchant to use Spellcheck" on menu!!!
???

**Comment by gedakc** [comment]
@Weej1 and @ScullyBlue, would you be able to try a new manuskript package I created?

The package was built using Debian 9 Stretch as the base OS and is called `manuskript-0.4.0-1-deb9-linux64.zip`.

Download location and instructions can be found on [SourceForge.Net - manuskript-0.4.0](https://sourceforge.net/projects/manuskript/files/manuskript/manuskript-0.4.0/).

**Comment by ScullyBlue** [comment]
It works for me! I downloaded the manuskript-0.4.0-1-deb9-linux64.zip, unpack and run it and it works fine. Theres no Warning for PyEnchant in Tools and the spellchecker is running.

**Comment by gedakc** [comment]
Thanks @ScullyBlue for testing `manuskript-0.4.0-1-deb9-linux64.zip` and confirming that it runs and the spellchecker works for you.

**Comment by gedakc** [comment]
Closing this issue because the new Linux packages available from the [wiki](https://github.com/olivierkes/manuskript/wiki) alternate manuskript download link do include the spell checker.

Note that the Windows package includes the spell checker too.
